node/src/components/consensus/highway_core/state.rs

```diff
@@ -705,6 +709,120 @@ impl<C: Context> State<C> {
+                }
+                let unit = self.unit(hash);
+                if let Some(eq_hash) = unit.panorama[eq_idx].correct() {
+                    if seen_by_endorsed(eq_hash) {
+                        continue;
+                    }
+                    match opt_naive_by_wunit {
+                        None => opt_naive_by_wunit = Some(eq_hash),
+                        Some(other_hash) => {
+                            if self.sees_correct(eq_hash, other_hash) {
+                                opt_naive_by_wunit = Some(eq_hash);
+                            } else if !self.sees_correct(other_hash, eq_hash) {
+                                return false;
+                            }
+                        }
+                    }
+                } else {
+                    to_visit.extend(
+                        unit.panorama
+                            .iter_correct_hashes()
+                            .filter(|hash| added_to_to_visit.insert(hash)),
+                    );
+                }
+            }
+        }
+        let naive_by_wunit = match opt_naive_by_wunit {
+            None => return true,
+            Some(naive_by_wunit) => naive_by_wunit,
+        };
+
+        // Iterate over all earlier units by wunit.creator, and find all forks by eq_idx they
+        // naively cite. If any of those forks are incompatible with naive_by_wunit, the LNC is
+        // violated.
+        let mut opt_pred_hash = wunit.panorama[wunit.creator].correct();
+        while let Some(pred_hash) = opt_pred_hash {
+            let pred_unit = self.unit(pred_hash);
+            let seen_by_endorsed = |hash| {
+                pred_unit
+                    .endorsed
+                    .iter()
+                    .any(|e_hash| self.sees(e_hash, hash))
+            };
+            let mut to_visit: Vec<_> = pred_unit.panorama.iter_correct_hashes().collect();
+            let mut added_to_to_visit: HashSet<_> = to_visit.iter().cloned().collect();
+            while let Some(hash) = to_visit.pop() {
+                if seen_by_endorsed(hash) {
+                    continue;
+                }
+                let unit = self.unit(hash);
+                if let Some(eq_hash) = unit.panorama[eq_idx].correct() {
+                    if seen_by_endorsed(eq_hash) {
+                        continue;
+                    }
+                    if !self.sees_correct(eq_hash, naive_by_wunit)
+                        && !self.sees_correct(naive_by_wunit, eq_hash)
+                    {
+                        return false;
+                    }
+                } else {
+                    to_visit.extend(
+                        unit.panorama
+                            .iter_correct_hashes()
+                            .filter(|hash| added_to_to_visit.insert(hash)),
+                    );
+                }
+            }
+            if !pred_unit.panorama[eq_idx].is_faulty() {
+                return true;
+            }
```

Comment:
Can't this be checked at the beginning of the `while` loop?

Reply:
No: We have to iterate over all of our own earlier units up to the last one that saw the equivocator as correct. But _including_ that last one, because it might still naively cite the single fork that it can see.

Reply:
Can you add a unit test that would catch that? I rearranged the code and put this check at the beginning of the loop and all tests passed.

Reply:
I'll work on it in the afternoon. :+1: